[P2P] Add support for -listen

fassadlr opened the pull request:
Closes https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/3041

zeptin commented on src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs:
```diff
@@ -148,7 +148,7 @@ public ConnectionManagerSettings(NodeSettings nodeSettings)
 
             this.BanTimeSeconds = config.GetOrDefault<int>("bantime", nodeSettings.Network.IsTest() ? DefaultMisbehavingBantimeSecondsTestnet : DefaultMisbehavingBantimeSeconds, this.logger);
-
+            this.Listen = config.GetOrDefault<bool>("listen", false, this.logger);
```
Doesn't this default to `true` for bitcoind?

zeptin replied:
`net.h`:

```
static const bool DEFAULT_LISTEN = true;
```

`init.cpp`:

```
    if (gArgs.IsArgSet("-bind")) {
        if (gArgs.SoftSetBoolArg("-listen", true))
            LogPrintf("%s: parameter interaction: -bind set -> setting -listen=1\n", __func__);
    }
    if (gArgs.IsArgSet("-whitebind")) {
        if (gArgs.SoftSetBoolArg("-listen", true))
            LogPrintf("%s: parameter interaction: -whitebind set -> setting -listen=1\n", __func__);
    }

    if (gArgs.IsArgSet("-connect")) {
        // when only connecting to trusted nodes, do not seed via DNS, or listen by default
        if (gArgs.SoftSetBoolArg("-dnsseed", false))
            LogPrintf("%s: parameter interaction: -connect set -> setting -dnsseed=0\n", __func__);
        if (gArgs.SoftSetBoolArg("-listen", false))
            LogPrintf("%s: parameter interaction: -connect set -> setting -listen=0\n", __func__);

...

    if (!gArgs.GetBoolArg("-listen", DEFAULT_LISTEN)) {
        // do not map ports or try to retrieve public IP when not listening (pointless)
        if (gArgs.SoftSetBoolArg("-upnp", false))
            LogPrintf("%s: parameter interaction: -listen=0 -> setting -upnp=0\n", __func__);
        if (gArgs.SoftSetBoolArg("-discover", false))
            LogPrintf("%s: parameter interaction: -listen=0 -> setting -discover=0\n", __func__);
        if (gArgs.SoftSetBoolArg("-listenonion", false))
            LogPrintf("%s: parameter interaction: -listen=0 -> setting -listenonion=0\n", __func__);
    }
```

dangershony replied:
This happens later
```
            if (this.Listen.Count == 0)
            {
                this.Listen.Add(new NodeServerEndpoint(new IPEndPoint(IPAddress.Parse("0.0.0.0"), this.Port), false));
            }
```

zeptin replied:
@dangershony that seems to be an entirely different `Listen` though? I am referring to the default value of the bool in the settings.

In any case it would be misleading to default it to `false` if code elsewhere treats it as `true` by default.

dangershony replied:
Ok in this case the listen is there just to override `-connect` perhaps the correct way is to explicitly set to false in case `-connect` it specified and `Listen` is not @fassadlr ?

fassadlr replied:
@zeptin thanks for the bitcoin core reference. I am going to implement it as they have done above.

zeptin commented on src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs:
```diff
@@ -182,6 +182,8 @@ public static void BuildDefaultConfigurationFile(StringBuilder builder, Network
             builder.AppendLine("####ConnectionManager Settings####");
             builder.AppendLine($"#The default network port to connect to. Default { network.DefaultPort }.");
             builder.AppendLine($"#port={network.DefaultPort}");
+            builder.AppendLine($"#Forces the node to accept incoming connections.");
```
Not sure 'forces' is really the intention, it's more that it allows them when `-listen` is set.

zeptin commented on src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs:
```diff
@@ -243,8 +246,17 @@ public static void PrintHelp(Network network)
+        /// <summary>
+        /// Forces the node to accept incoming connections.
+        /// <para>
+        /// This can be used to override the default behaviour when specifying peers to connect to
```
I suppose the word choice makes more sense in light of this.

fassadlr commented:
@zeptin @dangershony I have now implemented `-listen` as per bitcoin's specification. I've also added a `-forcelisten` argument which we can use to override the above behavior.

dangershony commented on src/Stratis.Bitcoin/Configuration/Settings/ConnectionManagerSettings.cs:
```diff
@@ -182,6 +180,10 @@ public static void BuildDefaultConfigurationFile(StringBuilder builder, Network
             builder.AppendLine($"#port={network.DefaultPort}");
+            builder.AppendLine($"#Accept connections from the outside.");
+            builder.AppendLine($"#listen=<0 or 1>");
+            builder.AppendLine($"#Forces the node to listen for incoming connections.");
```
maybe add in braces `(this can be used to accept incoming connections when -connect is specified)`

fassadlr replied:
Fixed 👍
